### PR TITLE
Use a filter to ensure that sessions have a user profile

### PR DESCRIPTION
**Pull request by yotommy**
### Description

Ensure that user-facing requests have a `pac4j` profile through a Play filter.

Currently, many of our controller methods check to see if there is a user profile in the session; if not, they invoke the `GuestClient` callback and redirect the request to store the profile ([example](https://github.com/civiform/civiform/blob/5decb99d8e5975d371e205fba40de3fb6a792bfe/server/app/controllers/applicant/ApplicantProgramReviewController.java#L89-L92)). This is tedious, error-prone, and brittle.

Instead, we check for the presence of a `pac4j` profile in the new `CiviFormProfileFilter`. If none is present, the callback and redirection are handled there.

Subsequent PRs will remove the special handling from controller methods.

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label: < feature | enhancement | bug | dependencies | infrastructure | ignore-for-release | database >
- [x] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers).
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)
- [ ] Extended the README / documentation, if necessary

### Issue(s) this completes

Relates to #5971

### CODE REVIEW

**Comment by yotommy** [comment]
* **#6809** <a href="https://app.graphite.dev/github/pr/civiform/civiform/6809?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @yotommy and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: server/app/filters/CiviFormProfileFilter.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  public CiviFormProfileFilter(Materializer mat, ProfileUtils profileUtils) {
+    super(mat);
+    this.profileUtils = checkNotNull(profileUtils);
+  }
+
+  private boolean profileIsMissing(Http.RequestHeader requestHeader) {
+    return profileUtils.currentUserProfile(requestHeader).isEmpty();
+  }
+
+  private boolean shouldApplyThisFilter(Http.RequestHeader requestHeader) {
```
Comment (gwendolyngoetz) [inline server/app/filters/CiviFormProfileFilter.java:33]: (blocking) There's a number of bool checks here. Can you add some javadoc to summarize the rules?

Comment (yotommy) [reply server/app/filters/CiviFormProfileFilter.java:40]: Done

**Review by gwendolyngoetz — APPROVED** [review]

**Review by dkatzz — APPROVED** [review]

File: server/app/filters/CiviFormProfileFilter.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import play.mvc.Http;
+import play.mvc.Result;
+
+/**
+ * Filter that ensures that sessions for user-facing requests have a CiviFormProfile.
+ *
+ * <p>The filter will redirect the request to the GuestClient callback under these conditions:
+ *
+ * <p>- The request is for a user-facing route - The request uses the `GET` method (since we
+ * redirect) - The session associated with the request does not contain a pac4j user profile
```
[the remark this answers is not in the document]
Comment (gwendolyngoetz) [reply server/app/filters/CiviFormProfileFilter.java:25]: @dkatzz If you mean the `<p>` tags it should. We live in a dystopian world where you don't have to close elements :cry:

[the remark this answers is not in the document]
Comment (dkatzz) [reply server/app/filters/CiviFormProfileFilter.java:25]: haha :) that did take some getting used to. But I was talking about the different bullets. Should we use <ul>?

[the remark this answers is not in the document]
Comment (yotommy) [reply server/app/filters/CiviFormProfileFilter.java:25]: Thanks for flagging the weird formatting. I had originally had these on separate lines, but `bin/fmt` wrapped them.

Used more conventional tags for the list items. Also tried to follow some online guidance I saw (don't close the list item tags, preface the list itself with a new paragraph).

**Pull request merged**